Faster time encoding for Handle Scribe

sopvop opened the pull request:
This implements time encoding functions using low-level Data.Test interface.

Encoding functions do not allocate anything except the result, and work really fast.

The Handle Scribe benchmark improved almost 4x times. From 24us to 6.1 us.

This should probably be used in JSON encoding, at least in the toJSON part.

The toEncoding part should use efficient implementation in Data.Aeson to avoid allocating Text or ByteString outside of Builder.

Fixes #3

Benchmark results before

```
benchmarking Katip.Scribes.Handle/Bytestring Builder
time                 23.85 μs   (23.41 μs .. 24.35 μs)
                     0.996 R²   (0.993 R² .. 0.999 R²)
mean                 24.11 μs   (23.68 μs .. 24.82 μs)
std dev              1.978 μs   (1.220 μs .. 2.994 μs)
variance introduced by outliers: 79% (severely inflated)
```

After

```
benchmarking Katip.Scribes.Handle/Bytestring Builder
time                 6.106 μs   (6.025 μs .. 6.223 μs)
                     0.997 R²   (0.995 R² .. 0.999 R²)
mean                 6.212 μs   (6.099 μs .. 6.524 μs)
std dev              569.2 ns   (258.0 ns .. 1.164 μs)
variance introduced by outliers: 85% (severely inflated)
```

We should add JSON benchmark and check if this may improve it also.

MichaelXavier commented:
@ozataman Any thoughts on this? The benchmark results look pretty nice. This change would not break any compatibility as its an internal change. Currently I am not aware of any plans to support custom time formats in the Handle scribe (of course it could be accomplished with a custom scribe), and ISO8601 seems like the best target format. I am a bit surprised there isn't a package out there that provides this formatting functionality. Maybe there is but it targets a different time library.

ozataman commented:
@sopvop Thank you for this work! I'm certainly not opposed to performance improvements, especially when they don't change anything at the interface level for the user. 

Lacking the time to go into it myself right now, here are the few things I'd want to ask to make sure:

- Currently, the output time format wouldn't change at all, right? I.e. no visible changes can be discerned by the user after this update?

- Any partial applications or anything like that that could raise unexpected exceptions due to this change?

- I've noticed there's already an associated test suite, which is great. Since this change relies on custom-made rendering functions, I guess it may be a good idea to take a quick look and see if we've covered all obvious cases. Common pitfalls in date related rendering typically are things like 0-padding output dates, etc. 

In general, our biggest fear is to be a source of exceptions or missed logs in production for the users of our library. Thanks again for the work @sopvop!

sopvop commented:
Btw, bikeshedding is very much welcome. If you don't like function or module names or anything.

I've added the ISO8601 variant so it can be used in json encoding part. Also both of these functions may be useful for implementing other scribes for katip.

--------

Not directly related to this PR:

Another possible improvement to Handle Scribe is encoding to utf-8 lazy bytestring before outpting it to handle.

Text.hPutBuf goes through GHC locale handling and iconv, even though most of computers use utf-8 based locales, at least on unixes.

Encoding to utf8 and using `D.B.Lazy.hPut` bench improves even more, shaving off another second.

```
benchmarking Katip.Scribes.Handle/Bytestring Builder
time                 4.840 μs   (4.775 μs .. 4.937 μs)
                     0.997 R²   (0.994 R² .. 0.999 R²)
mean                 4.891 μs   (4.823 μs .. 5.148 μs)
std dev              384.2 ns   (197.7 ns .. 757.9 ns)
variance introduced by outliers: 81% (severely inflated)
```
Changing existing handler like that is breaking BC, so we may instead introduce mkHandleScribeUtf8.

And in that case changing from Text.Builder to ByteString.Builder may improve things even more...

sopvop commented:
It should output exactly same format as before. I've added tests for that case.
The only difference should be years before 1000, those will be zero padded. And negative years will be absed.

There are no partial functions, the only thin place is allocating Array of proper length. And I've double-checked that. (you may want to tripple-check that, just to be sure)

Also this patch does not change the elasticsearch scribe.

sopvop commented:
It seems that the remaining build failure was introduced in 378bc3c44d0bcb4a0e83c7db1fd519ed2c969808

sopvop commented:
>  Currently, the output time format wouldn't change at all, right? I.e. no visible changes can be discerned by the user after this update?

In last commit I've added test against "golden" output by Handle scribe generated by original code.

>  Any partial applications or anything like that that could raise unexpected exceptions due to this change?

All functions are total, unlike formatter from aeson, which is partial with negative years.

>   I've noticed there's already an associated test suite, which is great. Since this change relies on custom-made rendering functions, I guess it may be a good idea to take a quick look and see if we've covered all obvious cases. Common pitfalls in date related rendering typically are things like 0-padding output dates, etc.

I've added several tests to check that output matches, if you think I've missed some corner cases, I'll be happy to add tests for them.

The only possible formatting change is zero padded years, and I don't think this is relevant for a logger.

MichaelXavier commented:
@sopvop Thanks for your hard work on this! Everything looks great. I'll probably be cutting a new release today. I'll add your name to the contributors list.

MichaelXavier commented:
Released in 0.3.1.0.
